src/Database/Expression/WhenThenExpression.php

```diff
@@ -0,0 +1,321 @@
+            throw new InvalidArgumentException(sprintf(
+                'The `$type` argument must be either an array, a string, or `null`, `%s` given.',
+                getTypeName($type)
+            ));
+        }
+
+        if (is_array($when)) {
+            if (
+                $type !== null &&
+                !is_array($type)
+            ) {
+                throw new InvalidArgumentException(sprintf(
+                    'When using an array for the `$when` argument, the `$type` argument must be an ' .
+                    'array too, `%s` given.',
+                    getTypeName($type)
+                ));
+            }
+
+            // avoid dirtying the type map for possible consecutive `when()` calls
+            $typeMap = clone $this->_typeMap;
+            if (
+                is_array($type) &&
+                count($type) > 0
+            ) {
+                $typeMap = $typeMap->setTypes($type);
+            }
+
+            $when = new QueryExpression($when, $typeMap);
+        } else {
+            if (
+                $type !== null &&
+                !is_string($type)
+            ) {
+                throw new InvalidArgumentException(sprintf(
+                    'When using a non-array value for the `$when` argument, the `$type` argument must ' .
+                    'be a string, `%s` given.',
+                    getTypeName($type)
+                ));
+            }
+
+            if ($type === null) {
+                $type = $this->_inferType($when);
+            }
+        }
+
+        $this->_when = $when;
+        $this->_whenType = $type;
+
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getWhenType()
+    {
+        return $this->_whenType;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getThen()
+    {
+        return $this->_then;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function then($result, ?string $type = null)
+    {
+        if (
+            $result !== null &&
+            !is_scalar($result) &&
+            !(is_object($result) && !($result instanceof Closure))
+        ) {
+            throw new InvalidArgumentException(sprintf(
+                'The `$result` argument must be either `null`, a scalar value, an object, ' .
+                'or an instance of `\%s`, `%s` given.',
+                ExpressionInterface::class,
+                getTypeName($result)
+            ));
+        }
+
+        $this->_then = $result;
+
+        if ($type === null) {
+            $type = $this->_inferType($result);
+        }
+
+        $this->_thenType = $type;
+
+        $this->_hasThenBeenDefined = true;
+
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getThenType(): ?string
+    {
+        return $this->_thenType;
+    }
+
+    /**
+     * Returns the type map.
+     *
+     * @return \Cake\Database\TypeMap
+     */
+    protected function getTypeMap(): TypeMap
```

Comment:
Do you need a method to access the property? Couldn't you use the property directly?

Reply:
Not necessarily needed, no. It is to satisfy `CaseExpressionTrait::inferType()`, which uses `getTypeMap()` since it's already available from the `TypeMapTrait`, which in turn was just used out of lazyness, and resulted in no type map being set by default for `CaseStatementExpression`, hence making me use `getTypeMap()` in the first place which creates a map if none is present 😆

The `CaseStatementExpression` constructor could certainly create a default type map, then `CaseExpressionTrait::inferType()` could use the property.